### PR TITLE
feat(api): reranking backend integrated in with rag

### PR DESCRIPTION
**Pull request by CollectiveUnicorn**
## Description

* Adds reranking to RAG pipeline
* Adds RAG configuration endpoint when in dev mode
* Adds additional logging
* Refactors the pytest's `test_routes` api tests
* Alters default RAG values into two steps, retrieval and ranking. With the retrieval results being set to 100 after ranking the results are filtered down to the user specified k value. If reranking is not enabled, the user specified k results is returned from the retrieval step.
* Adds Zarf configs to enable dev mode.

## FlashRank Evals
* Rerank (`ms-marco-TinyBERT-L-2-v2`) + 100 top-k retrieval

```
Final Results:
INFO:root:Average Needle in a Haystack (NIAH) Retrieval: 1.0
INFO:root:Average Needle in a Haystack (NIAH) Response: 1.0
INFO:root:Average Correctness (GEval): 0.812
INFO:root:Average Answer Relevancy: 0.9590000000000001
INFO:root:Average Annotation Relevancy: 0.92
INFO:root:MMLU: 0.697979797979798
INFO:root:HumanEval: 0.96
INFO:root:Eval Execution Runtime (seconds): 3122.107043981552
```

* Rerank (`ms-marco-TinyBERT-L-2-v2`) + 10 top-k retrieval
```
Final Results:
INFO:root:Average Needle in a Haystack (NIAH) Retrieval: 1.0
INFO:root:Average Needle in a Haystack (NIAH) Response: 1.0
INFO:root:Average Correctness (GEval): 0.8140000000000001
INFO:root:Average Answer Relevancy: 0.9470000000000001
INFO:root:Average Annotation Relevancy: 0.92
INFO:root:MMLU: 0.697979797979798
INFO:root:HumanEval: 0.96
INFO:root:Eval Execution Runtime (seconds): 3119.72075009346
```
* No Rerank
```
Final Results:
INFO:root:Average Needle in a Haystack (NIAH) Retrieval: 1.0
INFO:root:Average Needle in a Haystack (NIAH) Response: 1.0
INFO:root:Average Correctness (GEval): 0.8299999999999998
INFO:root:Average Answer Relevancy: 0.9555128205128205
INFO:root:Average Annotation Relevancy: 0.92
INFO:root:MMLU: 0.697979797979798
INFO:root:HumanEval: 0.96
INFO:root:Eval Execution Runtime (seconds): 3112.184923171997
```
* [Baseline](https://github.com/defenseunicorns/leapfrogai/pull/1078#issuecomment-2372599711) - No Rerank
```
Final Results:
INFO:root:Average Needle in a Haystack (NIAH) Retrieval: 1.0
INFO:root:Average Needle in a Haystack (NIAH) Response: 1.0
INFO:root:Average Correctness (GEval): 0.81
INFO:root:Average Answer Relevancy: 0.9616666666666666
INFO:root:Average Annotation Relevancy: 0.92
INFO:root:MMLU: 0.695959595959596
INFO:root:HumanEval: 0.96
INFO:root:Eval Execution Runtime (seconds): 1961.5159051418304
```

## Checklist before merging

- [x] Tests, documentation, ADR added or updated as needed
- [x] Followed the [Contributor Guide Steps](https://github.com/defenseunicorns/leapfrogai/blob/main/.github/CONTRIBUTING.md)

### CODE REVIEW

**Review by justinthelaw — APPROVED** [review]
The MVP implementation logic is sound to me - my only worry is a mix of a) our evals being skewed (too easy, too hard, for certain techniques, etc.) and, b) this re-ranking actually making our eval results worse with a noticable performance hit for the end-user. Not for this PR to solve, just my thoughts!

**Comment by CollectiveUnicorn** [comment]
Yeah the way the eval results are atm makes things a bit hard to interpret. But performance I think isn't as different as it seems. The baseline was from @jalling97's machine vs mine. The actual time difference between the version without reranking and with reranking is between 7 and 10 seconds over a 51 minute test. Would be better if we got an average per request but I suppose that could be evaluated separately.

**Comment by jalling97** [comment]
The `Annotation Relevancy` metric is what I was initially concerned with when I saw the results. I would have expected it to change (in either direction) after reranking was implemented. However, from what I can tell, the top_k chunks (with or without reranking) correspond to the same documents either way, resulting in no practical difference in annotations. Since the `Annotation Relevancy` metric isn't concerned with the order of the documents (just that the ones you expect are in the list of annotations),  the results are pretty much unchanged. 

This is good to note, as the next stage of evals will incorporate the chunk data thanks to #870, so we can better evaluate the rank of the chunks themselves. 

Overall I think this is a net gain and we'll have to see with the next wave of evals what practical difference the reranker has.

**Review by jalling97 — DISMISSED** [review]

File: tests/utils/client.py [hunk 1]
```diff
@@ -4,6 +4,10 @@
 import requests
 from requests import Response
 
+LEAPFROGAI_MODEL = os.getenv("LEAPFROGAI_MODEL", "llama-cpp-python")
+OPENAI_MODEL = "gpt-4o-mini"
```
Comment (justinthelaw) [inline tests/utils/client.py:8]: Similarly, this should use the `get_openai_model()` method on line 35.

File: tests/utils/client.py [hunk 2]
```diff
@@ -4,6 +4,10 @@
...
+ROOT_URL = "https://leapfrogai-api.uds.dev"
```
Comment (justinthelaw) [inline tests/utils/client.py:9]: Similarly, this should use the `get_leapfrogai_api_url()` method on line 64.

**Review by justinthelaw — CHANGES_REQUESTED** [review]

File: tests/utils/client.py (hunk 2 above)
Comment (CollectiveUnicorn) [reply tests/utils/client.py:9]: Can't use the `get_leapfrogai_api_url` because that's the full url with `openai/v1` at the end

Comment (justinthelaw) [reply tests/utils/client.py:9]: Sorry, my bad! I meant `get_leapfrogai_api_url_base()`

Comment (CollectiveUnicorn) [reply tests/utils/client.py:9]: I actually think its already switched, I just need to get rid of this.

**Review by jalling97 — DISMISSED** [review]

File: tests/integration/api/test_rag_files.py [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
+from typing import Optional
+
+import requests
 from openai.types.beta.threads.text import Text
 import pytest
 from tests.utils.data_path import data_path
 
-from tests.utils.client import client_config_factory
+from leapfrogai_api.typedef.rag.rag_types import ConfigurationPayload
+from tests.utils.client import ROOT_URL, client_config_factory
```
Comment (justinthelaw) [inline tests/integration/api/test_rag_files.py:10]: This import should be changed over to the `get_leapfrogai_api_url_base()` method

**Review by justinthelaw — APPROVED** [review]

**Review by jalling97 — APPROVED** [review]

**Pull request merged**